### ISSUE TITLE
The refresh token is not stored or used

### ISSUE THREAD

**Issue by benjumanji** [gndean/home-assistant-hypervolt-charger#56]
When logging in via direct grant with the offline_access scope a refresh token is returned with an infinite lifespan as long as it is used once every 30 days. Currently most HA users log in once every 5 minutes and create an offline session each time. Most users have around 2 to 3 offline sessions for various clients. People using home assistant have on average 8000 offline sessions running concurrently.

Suggestions
1. (Preferred) Store the refresh token, and when the access token expires after an hour, just use the refresh token to get a new access token. This does not create a new session. If for whatever reason the refresh grant fails you can always just start from scratch with a direct grant.
2. (easy) Remove the offline_access scope. Use the access token until it expires, then use a direct grant to get a new access token. This does not create a long lived session. It will expire quickly.

Which ever path is chosen it would be really nice if the access_token was at least reused until is expires or is at least somewhat close to expiration (i.e. at least to the half-life). It is far cheaper to check the signature on a JWT than it is to run 10s of thousands of hashes to check a password.

Refresh grants are just a simple post:
```
curl -XPOST https://kc.prod.hypervolt.co.uk/realms/retail-customers/protocol/openid-connect/token \
            -d grant_type=refresh_token \
            -d 'client_id=home-assistant' \
            -d refresh_token=$refresh_token
```
where `$refresh_token` is the refresh_token member of the JSON received from the `/token` endpoint.

**Comment by gndean** [comment]
I can look at implementing #1.

However, I think there's an even more preferred solution, and that's to use Home Assistant's built-in support for OAuth. The docs aren't super clear but it's referenced here:
https://developers.home-assistant.io/docs/config_entries_config_flow_handler#configuration-via-oauth2
and
https://developers.home-assistant.io/docs/api_lib_auth

This should allow HA to handle the refreshing of tokens and use a tried-and-tested framework rather than having to roll our own.

We discussed this back in 2023 and had the start of a POC. It's looks likely do-able but needs Hypervolt to support a callback for the HA client ID. Is that something you can support?

**Comment by gndean** [comment]
> Which ever path is chosen it would be really nice if the access_token was at least reused until is expires *or is at least somewhat close to expiration (i.e. at least to the half-life).*

Is there any way to know when the refresh token will expire?
The auth API appears to return
```"refresh_expires_in": 0,```
and the OAuth docs suggest that it's best practice not to know or assume any particular expiry time for the refresh token: https://www.oauth.com/oauth2-servers/making-authenticated-requests/refreshing-an-access-token/ (second to last paragraph)

**Comment by benjumanji** [comment]
> We discussed this back in 2023 and had the start of a POC. It's looks likely do-able but needs Hypervolt to support a callback for the HA client ID. Is that something you can support?

Looking at the docs, the only sticking point I can see is that they want to do code flow with a client secret. That is somewhat unfortunate because now I need to trust someone else (with whom I have no trust relationship) to keep that secret secret. It would be better if they were using PKCE (no confidentiality required). If you can:

a) supply the callback URL
b) get the home assistant side arranged to the point where there is _some_ secure channel over which I can share a client secret (ideally a public gpg key or similar) with them.
c) In big bold letters wherever someone is going to sign in with this that hypervolt does not warranty the confidentiality or security of any tokens supplied and use of this service is strictly at the users own risk.

Then ok.

**Comment by gndean** [comment]
I've released V2.2.5 which implements the 1st of the suggestions: it waits until the the websocket is closed or an API call fails with auth error, then attempts to refresh the access token. If that fails, it falls back to the previous login flow.

**Comment by gndean** [comment]
I'll close this for now as hopefully, over time, @benjumanji you'll see an improvement. I don't know how quickly HA users will adopt this new version.
I'll still like to revisit the HA supported OAuth2 flow, as could this help address #22, which I don't think I can resolve on my own.
